worker/credentialvalidator/worker.go

```diff
@@ -0,0 +1,162 @@
+// a cloud credential that a model uses.
+type Facade interface {
+
+	// ModelCredential gets model's cloud credential.
+	// Models that are on the clouds that do not require auth will return
+	// false to signify that credential was not set.
+	ModelCredential() (base.StoredCredential, bool, error)
+
+	// WatchCredential gets cloud credential watcher.
+	WatchCredential(string) (watcher.NotifyWatcher, error)
+}
+
+// IsValid returns true when the given credential is valid.
+func IsValid(c base.StoredCredential) bool {
+	return c.Valid
+}
+
+// Config holds the dependencies and configuration for a Worker.
+type Config struct {
+	Facade Facade
+	Check  Predicate
+}
+
+// Validate returns an error if the config cannot be expected to
+// drive a functional Worker.
+func (config Config) Validate() error {
+	if config.Facade == nil {
+		return errors.NotValidf("nil Facade")
+	}
+	if config.Check == nil {
+		return errors.NotValidf("nil Check")
+	}
+	return nil
+}
+
+// New returns a Worker that tracks the result of the configured
+// Check on the Model's cloud credential, as exposed by the Facade.
+func New(config Config) (*Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	modelCredential, exists, err := config.Facade.ModelCredential()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !exists {
+		return nil, ErrModelDoesNotNeedCredential
+	}
+	watcher, err := config.Facade.WatchCredential(modelCredential.CloudCredential)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Worker{
+		config:          config,
+		watcher:         watcher,
+		modelCredential: modelCredential,
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Init: []worker.Worker{watcher},
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Worker implements worker.Worker and util.Flag, and exits
+// with ErrChanged whenever the result of its configured Check of
+// the Model's cloud credential changes.
+type Worker struct {
+	catacomb        catacomb.Catacomb
+	config          Config
+	watcher         watcher.NotifyWatcher
+	modelCredential base.StoredCredential
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// Check is part of the util.Flag interface.
+func (w *Worker) Check() bool {
+	return w.config.Check(w.modelCredential)
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case _, ok := <-w.watcher.Changes():
+			if !ok {
+				return w.catacomb.ErrDying()
+			}
+			mc, exists, err := w.config.Facade.ModelCredential()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if !exists {
+				// Really should not happen in practice since the uninstall
```

Comment:
given the comment should we at least log that this has occurred as a Warning

Reply:
done